infix/parser.go

```diff
@@ -0,0 +1,224 @@
+package infix
+
+import (
+	"fmt"
+	"strconv"
+)
+
+type parser struct {
+	tokens chan Tokval
+	pocket *Tokval // lookahead
+}
+
+var TokEOF = Tokval{
+	T: EOF,
+	V: "<eof>",
+}
+
+func parserErr(tok Tokval) error {
+	return fmt.Errorf("unexpected %s", tok)
+}
+
+func Parse(code string) (Node, error) {
+	p := &parser{
+		tokens: Lex(code),
+	}
+	return p.parseExpr()
+}
+
+func (p *parser) peek() Tokval {
+	if p.pocket != nil {
+		val := *p.pocket
+		p.pocket = nil
+		return val
+	}
+
+	val, ok := <-p.tokens
+	if !ok {
+		return TokEOF
+	}
+	p.pocket = &val
+	return val
+}
+
+func (p *parser) next() Tokval {
+	if p.pocket != nil {
+		val := *p.pocket
+		p.pocket = nil
+		return val
+	}
+	tok, ok := <-p.tokens
+	if !ok {
+		return TokEOF
+	}
+	return tok
+}
+
+func (p *parser) parseExpr() (Node, error) {
+	var eoferr = func(expect string) error {
+		return fmt.Errorf("premature eof, expects %s",
```

Comment:
Perhaps "premature eof, expected %s" would be more clear (not sure, my english is far from great)

Reply:
hmm Maybe expected is better, indeed.